mono/mini/interp/interp.c

```diff
@@ -4047,7 +4047,7 @@ interp_exec_method_full (InterpFrame *frame, ThreadContext *context, guint16 *st
 			if (MONO_TYPE_IS_REFERENCE (m_class_get_byval_arg (c)))
 				mono_gc_wbarrier_generic_store (sp [-2].data.p, sp [-1].data.p);
 			else
-				stackval_from_data (m_class_get_byval_arg (c), sp [-2].data.p, (char *) &sp [-1].data.p, FALSE);
+				stackval_to_data (m_class_get_byval_arg (c), &sp [-1], sp [-2].data.p, FALSE);
```

Comment:
Shouldn’t it be &sp [-1].data.p?

Reply:
no, it shouldn't:

```c
static void inline
stackval_from_data (MonoType *type_, stackval *dst, char *src, gboolean pinvoke);

static void inline
stackval_to_data   (MonoType *type_, stackval *src, char *dst, gboolean pinvoke);
```

(The variable names are different in the actual source code, I changed them to make it more clear). `&sp [-1]` is of type `stackval *`. That thing before (`sp [-2].data.p`) was wrong, because it's of type `gpointer`. I'm a bit sad that the compiler didn't warn about it (or did we ignore it?). I had a look at the other `stackval_from_data` usages, they look fine.

They difference between those two functions is subtle, and in this specific case it _almost_ didn't matter. Let me elaborate what the problem here is: Assume we want to copy some `u2` value from the stack to an arbitrary address.

* previous case with `stackval_from_data`:
https://github.com/mono/mono/blob/ed15e75a943979a5b4d5e5dece8a9ae2cc4cc6cd/mono/mini/interp/interp.c#L459-L462

* now with `stackval_to_data`:
https://github.com/mono/mono/blob/ed15e75a943979a5b4d5e5dece8a9ae2cc4cc6cd/mono/mini/interp/interp.c#L542-L548

Note that `stackval` is defined like this:
https://github.com/mono/mono/blob/ed15e75a943979a5b4d5e5dece8a9ae2cc4cc6cd/mono/mini/interp/interp-internals.h#L54-L71

so before, as mentioned in the PR description, the destination address would write 4 bytes, even if we only want to copy 2 bytes. The previous case is only okay if the destination is _actually_ a value on the stack.

Reply:
Thanks for the explanation, I  just spotted the differences and wasn’t sure if it was a typo or on purpose.